src/lib.rs

```diff
@@ -240,59 +233,53 @@ impl Chip8 {
             0xB000 => self.pc = (0x0FFF & opcode) + self.registers[0] as u16,
             0xC000 => self.registers[index] = rand::random::<u8>() & kk, // random generator
             0xD000 => {
-                println!("TODO 0xD000");
                 // Draw a sprite, detecting collision
-                    /*
-                    uint8_t height = 0x000F & opcode;
-                    uint8_t x = Vx[indexX];
-                    uint8_t y = Vx[indexY];
-                    Vx[0xF] = 0;  // No collision to start with initially
+                let height = (0x000F & opcode) as u8;
+                let x = self.registers[index_x];
+                let y = self.registers[index_y];
+                self.registers[0xF] = 0; // No collision detected initially
 
-                    // Walk the sprite length (corresponding to height)
-                    for (int currentHeight = 0; currentHeight < height; ++currentHeight)
-                    {
-                        // Walk each sprite byte from MSB to LSB (corresponding to width)
-                        for (int currentWidth = 0; currentWidth < 8; ++currentWidth)
-                        {
-                            // Isolate the current bit
-                            uint8_t spriteByte = memory[I+currentHeight];
-                            uint8_t pixel = 0x80 & (spriteByte << currentWidth);
-                            pixel = pixel >> 7;
+                // Walk the length of the sprite (corresponding to height)
+                for current_height in 0..height {
+                    // Walk each sprite byte from MSB to LSB (corresponding to width)
+                    for current_width in 0..8 {
+                        // Isolate the current bit
+                        let sprite_byte: u8 =
+                            self.memory[(self.instruction_reg + current_height as u16) as usize];
+                        let mut pixel: u8 = 0x80 & (sprite_byte << current_width);
+                        pixel >>= 7;
 
-                            // Check for collision
-                            if (mScreen[(currentHeight+y)%32][(currentWidth+x)%64] & pixel)
-                            {
-                                Vx[0xF] = 1;
-                            }
-                            mScreen[(currentHeight+y)%32][(currentWidth+x)%64] ^= pixel;
+                        // Check for collision
+                        if self.screen[((current_height + y) % 32) as usize]
```

Comment:
You use this (Current_height + y) % 32 several times maybe just pull it out and do let y_offset = (Current_height + y) % 32.